File: 658-find-k-closest-elements/find-k-closest-elements.cpp
class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        vector<int> ans;
        int i=0;
        while(i < arr.size() && arr[i] < x) i++;
        int l = i-1, r = i;
        while(ans.size() < k){
            if(l >= 0 && r < arr.size()){
                if(abs(arr[l]-x) > abs(arr[r]-x)){
                    ans.push_back(arr[r]);
                    r++;
                } else {
                    ans.push_back(arr[l]);
                    l--;
                }
            } else if(l >= 0) {
                ans.push_back(arr[l]);
                l--;
            } else {
                ans.push_back(arr[r]);
                r++;
            }
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};

File: 658-find-k-closest-elements/find-k-closest-elements.py
class Solution:
    def findClosestElements(self, arr: List[int], k: int, x: int) -> List[int]:

        # O(nlgn) solution
        # diff_arr = []
        # for ele in arr:
        #     diff_arr.append([abs(x-ele),ele])
        
        # diff_arr.sort(key=lambda x: [x[0],x[1]])

        # result = [ele[1] for ele in diff_arr]
        # result.sort()
        # return result[:k]

        # O(n) solution
        low = 0
        high = len(arr)-1
        while((high-low) >= k):
            if abs(arr[low]-x) > abs(arr[high]-x):
                low += 1
            else:
                high -= 1
        
        result = []
        while(low <= high):
            result.append(arr[low])
            low += 1
        return result
        